### ISSUE TITLE
Proposal: no-optional-catch-binding rule

### ISSUE THREAD

**Issue by cjihrig** [eslint/eslint#10264]
**Please describe what the rule should do:**

Prevent use of the new optional `catch` binding syntax.

**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
try {
  throw new Error('error that should not be ignored');
} catch {
  console.log('something happened');
}
```

**Why should this rule be included in ESLint (instead of a plugin)?** It seems like it would be applicable across all environments that support the new syntax. Generally, errors should not be ignored. If an error is ignored, I'd prefer an explicit opt eslint-disable opt out, rather than not knowing if it was ignored accidentally or lazily.

I'm volunteering to do the work. I have a [proof of concept](https://github.com/cjihrig/no-optional-catch-binding) already.

**Comment by mysticatea** [comment]
Thank you for this proposal.

I like this rule.

We cannot accept this proposal for right now because we don't add rules for Stage 3 syntax to core in our policy (and ESTree spec has not exist yet). But I hope to add this rule to core in the future.

**Comment by cjihrig** [comment]
I was hesitant to open this issue because it's still Stage 3, but figured it might be worth a shot now that the syntax is in Node 10. I'm happy to wait until it hits Stage 4. Thanks for the feedback.

**Comment by platinumazure** [comment]
With optional catch bindings now in Stage 4, I think we could re-evaluate.

**Comment by ljharb** [comment]
Why should this be a rule?? It's *perfectly fine* to ignore errors in many cases, and choosing to omit the catch binding is explicit developer intent to ignore it.

For all the times eslint core members push back on adding reasonable rules with "just use no-restricted-syntax!", it seems like this is a PRIME candidate for "just use no-restricted-syntax".

**Comment by mysticatea** [comment]
We can develop this after #10392 which adds the support of optional `catch` binding is merged.

I will champion this.

- Generally, errors should not be ignored in order to investigate bugs fastly. Ignoring errors would help to hide the cause of bugs.
- The `no-restricted-syntax` rule with `CatchClause[param=null]` can catch this. However, it's inconvenient if sharable config is using the `no-restricted-syntax` rule. It overrides and mess the configuration of shareble configs. Independent rule is useful.

**Comment by ljharb** [comment]
@mysticatea if "it's inconvenient if sharable config is using the no-restricted-syntax rule" is going to apply here, then I'd respectfully request that *never again* is "use no-restricted-syntax" offered as a reason to reject a rule.

**Comment by mysticatea** [comment]
@ljharb I mentioned the reason of inconvenient.

**Comment by ljharb** [comment]
@mysticatea right, i 1000% agree that it's inconvenient :-) but that reasoning needs to either apply to every rule request, or none of them, otherwise the criteria is inconsistent.

**Comment by not-an-aardvark** [comment]
I agree with `ljharb`, this rule seems like a stylistic preference to me. I think the argument of "errors generally should not be ignored" would have been more compelling if it was made on the TC39 proposal before the proposal reached stage 4. This isn't a case where the language has a legacy feature that needs to remain for compatibility reasons -- it's a new feature which (at least according to TC39) presents a compelling addition to the language. Given that, I'm unconvinced that we should be encouraging people not to use it already.

**Comment by ljharb** [comment]
It was made, and TC39 dismissed it - because ignoring errors is *perfectly legitimate*.

**Comment by mysticatea** [comment]
At least, I think that it's very useful syntax in order to overlook the cause of bugs in my 2 cents.

**Comment by mysticatea** [comment]
I'm sorry if my comment is offensive. But I think it's enough general matter for core rules.

**Comment by ljharb** [comment]
`let foo; try { foo = JSON.parse(data); } catch {}` is 100% legitimate.

**Comment by mysticatea** [comment]
@ljharb It works well, but it abandons the information how it fails parsing, where the syntax error exists.

**Comment by ljharb** [comment]
There are many cases where that information is not actionable, and not useful (and to the developer, visible in the Network tab if they choose to debug it in a browser).

There's also no rule that prohibits argumentless `.catch` callbacks, because that's equally legitimate. Separately, `no-unused-vars` has special handling for unused catch bindings, for this very purpose.

**Comment by mysticatea** [comment]
I proposed `prefer-optional-catch-binding` (#10434) rule which recommends using optional `catch` binding if the error is ignored.

We can enforce the following style with the combination of those rules:

- Basically, don't ignore errors.
- But use optional `catch` binding if ignoring errors is reasonable, by explicitly opt-out.

**Comment by mysticatea** [comment]
One important thing, ESLint will lose the ability to enforce the style "Basically, don't ignore errors" by the new syntax. Now we can enforce to use `catch` bindings by the `no-unused-vars` rule. However, the `no-unused-vars` rule overlooks the optional `catch` binding syntax because the syntax doesn't define any variable. This rule is needed to keep the ability.

**Comment by not-an-aardvark** [comment]
There are a lot of ways to get around linting rules. A rule generally provides value by alerting developers about something they might have forgotten, not by covering every possible way that a dedicated user might try to get around it. In this case, I think `no-unused-vars` can still be useful for enforcing that errors aren't ignored, even if it doesn't guarantee that every error is handled in the way that one might intend.

**Comment by geek** [comment]
I'd prefer to never have an error ignored unless it was very explicit with a disable no-option-catch comment. A world that blatantly ignores errors without clear reasoning is really terrifying.

**Comment by ljharb** [comment]
The use of an optional catch binding *is* a very explicit indicator that one doesn't care *what* the error is, only that there was an error - which is, i'll add, not at all "ignoring errors". It's a catch block - it's impossible to ignore the error; only to ignore the specific value that was thrown.

**Comment by mysticatea** [comment]
@ljharb "ignore" means "abandoning the error information". There are people that want to disallow 
"one doesn't care what the error is" basically.

**Comment by mysticatea** [comment]
Because, the thrown error can be an unintentional error. For example, `let foo; try { foo = JSON.parse(e.data); } catch {}`, one intended to ignore parsing error, but actually it hided a TypeError if `e` is null. The `try` syntax has very wide effect.

**Comment by platinumazure** [comment]
Maybe it would be best to make a rule that could go either way, so people can use it how they like? I don't think optional catch bindings are so wrong that we should have an opinionated rule which only bans then. I like the idea of preferring either no catch binding (if the binding is unused) or requiring a binding in all cases.

**Comment by mysticatea** [comment]
@platinumazure Do you mean to merge this and #10434? Actually, I'd like to enable both in my case. I want to report "ignoring errors" basically, but I want to remove `catch` binding if I explicitly opt-outed the error with `//eslint-disable-line no-optional-catch-binding` comment for some reason.

**Comment by ljharb** [comment]
It seems like they should maybe just be the same rule.

**Comment by cjihrig** [comment]
I think it's a reasonable compromise to support both cases. I can update my code if there is agreement on what it should look like. How is this:

name: `prefer-optional-catch-binding`
values: `'always'`, `'never'`

When set to `'never'`, optional catch bindings would be forbidden, as described above. When set to `'always'`, it would enforce the usage of optional catch bindings and could be autofixed as described in #10434.

**Comment by not-an-aardvark** [comment]
To me, it still seems like the `always` behavior is redundant with `no-unused-vars`. I'm not sure it makes sense to add a new rule for that behavior just to support autofixing.

It's also not clear to me that a user would expect an autofixer to remove their catch binding. Often, when a variable is unused, the problem is that the user forgot to use it, not that they mistakenly included the declaration. (This is why we don't have an autofix for `no-unused-vars`.)

As a sidenote, if we do add a rule I think we should avoid names like `no-optional-catch-binding` and instead use names along the lines of `no-unused-catch-binding` or `prefer-catch-binding`. I think the term "optional" made sense as a name for the language proposal since contrasted with the previous state of the language (where catch bindings were required). Since catch bindings are always optional now, it seems redundant and potentially confusing to call them "optional catch bindings" in a rule name.

**Comment by ljharb** [comment]
It’s not redundant because no-unused-vars can’t autofix to optional catch, and because it’s purpose isn’t the same as this one, which deals specifically with catch clauses.

I don’t want to autofix unused vars, but i absolutely want to autofix catch clauses that could omit the binding.

**Comment by mysticatea** [comment]
> As a sidenote, if we do add a rule I think we should avoid names like no-optional-catch-binding and instead use names along the lines of no-unused-catch-binding or prefer-catch-binding. I think the term "optional" made sense as a name for the language proposal since contrasted with the previous state of the language (where catch bindings were required). Since catch bindings are always optional now, it seems redundant and potentially confusing to call them "optional catch bindings" in a rule name.

I agree. The purpose of this rule is to require to use of `catch` bindings or explicitly opt-out, so I think `require-catch-binding` is a better name. `// eslint-disable-line require-catch-binding` will be an understandable note for opt-out. If it's `// eslint-disable-line prefer-optional-catch-binding` then I feel odd.

----

To support the opposite side, I think better if it's a separated rule. Maybe `prefer-no-unused-catch-binding` or something like. The rule will overlap with `no-unused-vars` rule, but I think nice if core rules include the rules which encourage to learn new syntax.

**Comment by platinumazure** [comment]
@eslint/eslint-team Where are we at on this?

Do we need a new rule here (at least, for right now), when `no-restricted-syntax` could theoretically be used by people who want to disallow catch blocks with no bindings? Wondering if this is something for which we should wait for feedback before doing anything in core rules.

**Comment by mysticatea** [comment]
@platinumazure My opinion about `no-restricted-syntax` is here: https://github.com/eslint/eslint/issues/10264#issuecomment-394037014. Plus, I think valuable if there is the page which explains `try-catch` syntax catches all exceptions (it can include unintentional exceptions) as a rule.

**Comment by michaelficarra** [comment]
A new rule will make it seem like this is a reasonable thing to do, and I strongly believe that a rule like this should never be enabled. :-1: Do it with `no-restricted-syntax` if you must.

**Comment by ljharb** [comment]
> is not proper for the explicitly opt-out.

Restating my comment above: I agree with this in general, which is why I think "you can just use no-restricted-syntax" is and should be an invalid reason not to accept a rule.

> A new rule will make it seem like this is a reasonable thing to do,

However, I also agree with this, and think this shouldn't be a rule, because the whole point of this language feature is when you want error information disposed.

> to disallow disposal of error information 

If this is the rule you want, then you'd have to block other patterns too - including a catch with a binding that's unused; any try/finally that throws in the finally (thus disregarding any previously thrown exception); any unreturned/unpassed promise without a `.catch`; any promise with a `.finally` that throws or rejects in the finally (thus disregarding any previous rejection)… Otherwise, you're not actually achieving your goal.

**Comment by not-an-aardvark** [comment]
Re: "If this is the rule you want, then you'd have to block other patterns too", most of these patterns are already possible to block.

> a catch with a binding that's unused

This can be done with `no-unused-vars`.

> any try/finally that throws in the finally

This can be done with `no-unsafe-finally`.

> any promise with a .finally that throws or rejects in the finally (thus disregarding any previous rejection)

I don't think there's a built-in rule for this, although it's inherently a bit more difficult to detect statically since it's not a syntax construct.

**Comment by ljharb** [comment]
Sure. But if "disallow error disposal" needs a standalone rule, to be more explicit and clear (which in general makes sense to me) - then shouldn't all those things be blocked by this new rule?

**Comment by cjihrig** [comment]
One small note - I had proposed changing this to also incorporate #10434, if that makes any difference.

**Comment by mysticatea** [comment]
@ljharb Looks reasonable proposition. Yes, "disallow error disposal" rule is nice. However, currently, I think that each core rule is smaller than it. For example, if I want semicolon-less style, I need the following rules:

```json
{
    "semi": ["error", "never", {"beforeStatementContinuationChars": "never"}],
    "semi-spacing": ["error", {"after": true, "before": false}],
    "semi-style": ["error", "first"],
    "no-extra-semi": "error",
    "no-unexpected-multiline": "error"
}
```

The combination of small rules lets me get the goal. Each rule might be too small, but I think it's another issue that consideration of the proper rule size.

-----

Currently, I'm considering the following rule:

```js
{
    "require-catch-binding": ["error", "always" | "never"]
}
```

**Options:**

- `"always"` ... enforce writing `catch` binding.
- `"never"` ... enfoce omitting `catch` binding if it's unused. This supports autofix.

**Comment by Mouvedia** [comment]
If it's not added to recommended I am all for it.

**Comment by teddy-error** [comment]
The Optional catch binding proposal has been moved to stage 4

**Comment by kaicataldo** [comment]
This proposal has been open for quite a long time now. @mysticatea Are you still championing this? There's been a lot of discussion, and it would be great to have a clear proposal that the team can vote on.

**Comment by silverwind** [comment]
I'm currently "emulating" a requirement for optional catch binding via the [caughtErrors](https://eslint.org/docs/rules/no-unused-vars#caughterrors) option. I guess it could be a viable alternative, but a explicit rule for it might be even better.

**Comment by kaicataldo** [comment]
Since the creation of this proposal, we have started recommending using `no-restricted-syntax` for anything that it can reliably catch - especially when it's a simple rule like this. As a result, I'm now of the mind that we should defer to `no-restricted-syntax` to enforce this.

For anyone wanting to enforce this, you can use the following the configuration:

```js
/* eslint no-restricted-syntax: ["error", {
    "selector": "CatchClause[param=null]",
    "message": "Optional catch binding are not allowed."
}] */

try {
  throw new Error('error that should not be ignored');
} catch {
  console.log('something happened');
}
```

You can see it in action [here](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCB7XG4gICAgXCJzZWxlY3RvclwiOiBcIkNhdGNoQ2xhdXNlW3BhcmFtPW51bGxdXCIsXG4gICAgXCJtZXNzYWdlXCI6IFwiT3B0aW9uYWwgY2F0Y2ggYmluZGluZyBhcmUgbm90IGFsbG93ZWQuXCJcbn1dICovXG5cbnRyeSB7XG4gIHRocm93IG5ldyBFcnJvcignZXJyb3IgdGhhdCBzaG91bGQgbm90IGJlIGlnbm9yZWQnKTtcbn0gY2F0Y2gge1xuICBjb25zb2xlLmxvZygnc29tZXRoaW5nIGhhcHBlbmVkJyk7XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMCwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjYWxsYmFjay1yZXR1cm4iOjJ9LCJlbnYiOnt9fX0=).

**Comment by silverwind** [comment]
@kaicataldo I guess most users would actually want to require the optional catch binding, is that possible via `no-restricted-syntax` too?

**Comment by kaicataldo** [comment]
That's a good point, and yes! 

```js
/* eslint no-restricted-syntax: ["error", {
    "selector": "CatchClause[param!=null]",
    "message": "Optional catch bindings are required."
}] */

try {
  throw new Error('error that should not be ignored');
} catch (e) {
  console.log('something happened');
}
```

Demo example [here](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCB7XG4gICAgXCJzZWxlY3RvclwiOiBcIkNhdGNoQ2xhdXNlW3BhcmFtIT1udWxsXVwiLFxuICAgIFwibWVzc2FnZVwiOiBcIk9wdGlvbmFsIGNhdGNoIGJpbmRpbmdzIGFyZSByZXF1aXJlZC5cIlxufV0gKi9cblxudHJ5IHtcbiAgdGhyb3cgbmV3IEVycm9yKCdlcnJvciB0aGF0IHNob3VsZCBub3QgYmUgaWdub3JlZCcpO1xufSBjYXRjaCAoZSkge1xuICBjb25zb2xlLmxvZygnc29tZXRoaW5nIGhhcHBlbmVkJyk7XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMCwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjYWxsYmFjay1yZXR1cm4iOjJ9LCJlbnYiOnt9fX0=).

**Comment by silverwind** [comment]
I don't think it's really useful that way when one wants to [use the error](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCB7XG4gICAgXCJzZWxlY3RvclwiOiBcIkNhdGNoQ2xhdXNlW3BhcmFtIT1udWxsXVwiLFxuICAgIFwibWVzc2FnZVwiOiBcIk9wdGlvbmFsIGNhdGNoIGJpbmRpbmdzIGFyZSByZXF1aXJlZC5cIlxufV0gKi9cblxudHJ5IHtcbiAgdGhyb3cgbmV3IEVycm9yKCdlcnJvciB0aGF0IHNob3VsZCBiZSBpZ25vcmVkJyk7XG59IGNhdGNoIChlcnIpIHtcbiAgY29uc29sZS5sb2coZXJyKTtcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjEwLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImNhbGxiYWNrLXJldHVybiI6Mn0sImVudiI6e319fQ==). That said, I'm happy with the `caughtErrors` option of `no-unused-vars` personally. Maybe it should just be made default in a future version of ESLint.

**Comment by kaicataldo** [comment]
> @kaicataldo I guess most users would actually want to require the optional catch binding, is that possible via no-restricted-syntax too?

Did I misunderstand this comment? I thought you were asking if you could do the above.

**Comment by silverwind** [comment]
Well, you answered the exact question, but I guess I should have been more clear about the "optional" part, otherwise the rule would not really be practical.

 I think I'm now currently leaning to not implementing a new rule because:

1. `catch()` is invalid syntax which makes `no-unused-vars` usable for requiring optional bindings
2. `no-restricted-syntax` or [this plugin](https://eslint-plugin-es.mysticatea.dev/rules/no-optional-catch-binding.html) can be used for forbidding optional bindings

**Comment by kaicataldo** [comment]
Just to be clear, `catch() {}` is a parser error, and `CatchClause[param!=null]` represents `catch {}`.

**Comment by nzakas** [comment]
As already discussed, and as [announced](https://eslint.org/blog/2020/05/changes-to-rules-policies), we are no longer adding new core rules that simply disallow syntax. It's possible to use `no-restricted-syntax`, or to create your own rule to focus on just blocking this pattern.

Thanks for the discussion, everyone.
